### PR TITLE
Add ASR-EOU models

### PR DESCRIPTION
**Pull request by stevehuang52**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add ASR-EOU models

**Collection**: [asr]

# Changelog

- Add model, dataset, training/evaluation scripts, as well as utils scripts

### CODE REVIEW

File: nemo/collections/asr/modules/lstm_decoder.py [hunk 1]
```diff
@@ -57,7 +66,7 @@ def __init__(self, feat_in, num_classes, lstm_hidden_size, vocabulary=None, bidi
             self.__vocabulary = vocabulary
         self._feat_in = feat_in
         # Add 1 for blank char
-        self._num_classes = num_classes + 1
+        self._num_classes = num_classes + 1 if add_blank else num_classes
```
Comment (stevehuang52) [inline nemo/collections/asr/modules/lstm_decoder.py:70]: to make it as a general frame-level classifier

File: nemo/collections/asr/modules/ssl_modules/multi_layer_feat.py [hunk 2]
```diff
@@ -81,85 +84,12 @@ def forward(
             raise ValueError(f"Unknown mode {self.mode}")
 
 
-class ConformerMultiLayerFeatureExtractor(NeuralModule, Exportable):
```
Comment (stevehuang52) [inline nemo/collections/asr/modules/ssl_modules/multi_layer_feat.py:84]: remove to use the one from conformer_encoder.py to avoid duplicates

File: nemo/core/classes/modelPT.py [hunk 3]
```diff
@@ -1465,7 +1465,7 @@ def maybe_init_from_pretrained_checkpoint(self, cfg: OmegaConf, map_location: st
                 if isinstance(cfg.init_from_ptl_ckpt, str):
                     # Restore checkpoint
                     ckpt_path = cfg.pop('init_from_ptl_ckpt')
-                    ckpt = torch.load(ckpt_path, map_location=map_location)
+                    ckpt = torch.load(ckpt_path, map_location=map_location, weights_only=False)
```
Comment (stevehuang52) [inline nemo/core/classes/modelPT.py]: fix to avoid loading error

File: nemo/collections/asr/modules/rnnt.py [hunk 4]
```diff
@@ -1479,6 +1482,7 @@ def forward(
         transcripts: Optional[torch.Tensor] = None,
         transcript_lengths: Optional[torch.Tensor] = None,
         compute_wer: bool = False,
+        keep_hypotheses: bool = False,
```
Comment (stevehuang52) [inline nemo/collections/asr/modules/rnnt.py:1486]: add the option to maintain hypothesis for later usage, e.g., in calculating EOU metrics

File: nemo/collections/asr/losses/ssl_losses/mlm.py [hunk 5]
```diff
@@ -65,11 +65,14 @@ def forward(
         if masks is None:
             masks = spec_masks
 
-        # B,D,T -> B,T,D
-        masks = masks.transpose(1, 2)
+        if masks is None:
```
Comment (stevehuang52) [inline nemo/collections/asr/losses/ssl_losses/mlm.py:68]: fix bug when masks is None

File: scripts/speech_recognition/convert_to_tarred_audio_dataset.py [hunk 6]
```diff
@@ -563,15 +563,42 @@ def create_concatenated_dataset(
         metadata_yaml = OmegaConf.structured(metadata)
         OmegaConf.save(metadata_yaml, new_metadata_path, resolve=True)
 
-    def _read_manifest(self, manifest_path: str, config: ASRTarredDatasetConfig):
+    def _read_manifest(self, manifest_path: Union[str, List[str]], config: ASRTarredDatasetConfig):
```
Comment (stevehuang52) [inline scripts/speech_recognition/convert_to_tarred_audio_dataset.py:566]: add the option to read from multiple files

**Comment by nithinraok** [comment]
@stevehuang52 when you are back, probably we can prioritize this PR?

File: scripts/asr_eou/transcribe_speech_sharded.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+from nemo.collections.asr.modules.conformer_encoder import ConformerChangeConfig
+from nemo.collections.asr.parts.utils.manifest_utils import read_manifest, write_manifest
+from nemo.core.config import hydra_runner
+from nemo.utils import logging
+
+"""
+Transcribe audio manifests on distributed GPUs. Useful for transcription of moderate amounts of audio data.
+This script also supports splitting the manifest into chunks and merging the results back together.
+This script is a modified version of `transcribe_speech_distributed.py` that only takes manifest files as input.
+It is useful for transcribing a large amount of audio data that does not fit into a single job.
```
Comment (stevehuang52) [inline scripts/asr_eou/transcribe_speech_sharded.py:43]: @nithinraok Could you please take a look at this file to see if it's useful? I'm okay removing this file, but personally I have found it useful to transcribe a huge amount of audios that don't fit into a single cluster job.

Comment (nithinraok) [reply scripts/asr_eou/transcribe_speech_sharded.py:43]: Thanks steve. Can you check if we can merge https://github.com/NVIDIA-NeMo/NeMo/blob/main/examples/asr/transcribe_speech_parallel.py and this one.

File: nemo/collections/asr/metrics/wer.py [hunk 8]
```diff
@@ -255,7 +256,7 @@ def __init__(
         batch_dim_index=0,
         dist_sync_on_step=False,
         sync_on_compute=True,
-        **kwargs,
+        return_hypotheses=False,
```
Comment (nithinraok) [inline nemo/collections/asr/metrics/wer.py:260]: update docstrings

File: nemo/collections/asr/modules/rnnt.py (hunk 4 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/rnnt.py:1486]: update docstrings

File: nemo/collections/common/data/lhotse/dataloader.py [hunk 9]
```diff
@@ -541,6 +541,19 @@ def get_lhotse_sampler_from_config(config, global_rank, world_size, tokenizer=No
...
+            cut.is_mixed_noise = True
+            return cut
+
+        # In current lhotse implementation, if padding is applied before noise augmentation,
+        # and your noise manifest has dummy text field like `"text": "-"`,
+        # the call to MixCut.first_non_padding_cut will return the noise cut
+        # instead of the speech cut. We mark the noise cut with ``is_mixed_noise`` flag
+        # to avoid this issue, and the speech cut can be obtained by:
+        # `cut =[t.cut for t in cut.tracks if len(t.cut.supervisions) > 0 and not t.cut.custom.get("is_mixed_noise", False)][0]`
+        noise = noise.map(mark_as_mixed_in_noise)
```
Comment (nithinraok) [inline nemo/collections/common/data/lhotse/dataloader.py]: @pzelasko

File: scripts/asr_eou/tokenizers/add_special_tokens_to_sentencepiece.py
Comment (nithinraok) [inline scripts/asr_eou/tokenizers/add_special_tokens_to_sentencepiece.py:1]: should we keep this here or https://github.com/NVIDIA-NeMo/NeMo/tree/main/scripts/tokenizers

Comment (stevehuang52) [reply scripts/asr_eou/tokenizers/add_special_tokens_to_sentencepiece.py:1]: I hardcoded some special tokens for EOU so I feel it's better to keep it in a EOU related folder

File: scripts/asr_eou/transcribe_speech_sharded.py (hunk 7 above)
Comment (stevehuang52) [reply scripts/asr_eou/transcribe_speech_sharded.py:43]: The `transcribe_speech_parallel.py` assumes that all the audios can be processed in one job, thus it lacks the feature to split large manifests and resume only on unprocessed chunks. 

Personally, I feel that it's okay to drop `transcribe_speech_parallel.py` and use this new script which directly extends `transcribe_speech.py`, which makes future maintenance easier since any improvements on `transcribe_speech.py` will be automatically compatible with the distributed version. Currently, the features of `trancribe_speech.py` and `transcribe_speech_parallel.py` have diverged a lot, for example in `trancribe_speech.py`  we sort audios by duration, but it's not supported in `transcribe_speech_parallel.py`.

File: nemo/collections/asr/metrics/wer.py (hunk 8 above)
Comment (stevehuang52) [reply nemo/collections/asr/metrics/wer.py:260]: added

File: nemo/collections/asr/modules/rnnt.py (hunk 4 above)
Comment (stevehuang52) [reply nemo/collections/asr/modules/rnnt.py:1486]: added

File: examples/asr/transcribe_speech_manifest_distributed.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+
+from nemo.collections.asr.modules.conformer_encoder import ConformerChangeConfig
+from nemo.collections.asr.parts.utils.manifest_utils import read_manifest, write_manifest
+from nemo.core.config import hydra_runner
+from nemo.utils import logging
+
+"""
+Transcribe audio manifests on distributed GPUs. Useful for transcription of moderate amounts of audio data.
+This script also supports splitting the manifest into chunks and merging the results back together.
+This script is a modified version of `transcribe_speech_distributed.py` that only takes manifest files as input.
```
Comment (nithinraok) [inline examples/asr/transcribe_speech_manifest_distributed.py]: Please fix this  and rename this file to `transcribe_speech_distributed.py`

File: examples/asr/transcribe_speech_manifest_distributed.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+It is useful for transcribing a large amount of audio data that does not fit into a single job.
```
Comment (nithinraok) [reply examples/asr/transcribe_speech_manifest_distributed.py]: Also remove `parallel.py` script

File: examples/asr/transcribe_speech_manifest_distributed.py [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+
+    return all_sharded_manifest_files
+
+
+def maybe_merge_manifest(cfg: TranscriptionConfig):
+    if cfg.split_size is None or cfg.split_size <= 0:
+        return
+
+    # only merge manifest on the first GPU of the first node
+    if not cfg.gpu_idx == 0 and cfg.node_idx == 0:
```
Comment (nithinraok) [inline examples/asr/transcribe_speech_manifest_distributed.py]: should this be: `if not (cfg.gpu_idx == 0 and cfg.node_idx == 0)`

File: scripts/asr_eou/transcribe_speech_sharded.py (hunk 7 above)
Comment (nithinraok) [reply scripts/asr_eou/transcribe_speech_sharded.py:43]: does this support tarred dataset?

Comment (stevehuang52) [reply scripts/asr_eou/transcribe_speech_sharded.py:43]: Since `transcribe_speech.py` doesn't support tarred dataset, this one doesn't either.

File: examples/asr/transcribe_speech_manifest_distributed.py (hunk 12 above)
Comment (stevehuang52) [reply examples/asr/transcribe_speech_manifest_distributed.py]: you're right, fixed

File: scripts/asr_eou/transcribe_speech_sharded.py (hunk 7 above)
Comment (nithinraok) [reply scripts/asr_eou/transcribe_speech_sharded.py:43]: I am afraid then to replace parallel script as we use it on tarred sets like Granary for inference. Is there a way to support tarred set inference

Comment (stevehuang52) [reply scripts/asr_eou/transcribe_speech_sharded.py:43]: Yes I think so. How about we keep both for now, and then we work on a merged version later?

**Comment by stevehuang52** [comment]
Hi @nithinraok @tango4j @weiqingw4ng, I have listed the main changes of this PR in the PR description, let's work on each section one by one and hope that help ease your burden. Please note that the \~7k line addition has \~3k lines of yaml config, so the actual work is not as hard as it may seem at the first sight. If you prefer, feel free to assign yourself to different sections of the changes so we can split the workload. Thanks for reviewing this PR~!

File: examples/asr/asr_eou/README.md [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+
+We expect the input data manifest to be JSONL format, with each line containing the following fields:
+```json
+{
+    "audio_filepath": "/path/to/audio.wav",
+    "text": "The text of the audio.", # transcript of the utterance
+    "offset": 0.0,  # offset of the audio, in seconds
+    "duration": 3.0,  # duration of the audio, in seconds
+    "sou_time": 0.2,  # start of utterance time, in seconds
+    "eou_time": 1.5,  # end of utterance time, in seconds
```
Comment (pzelasko) [inline examples/asr/asr_eou/README.md:90]: could have been `MonoCut(..., duration=3.0, supervisions=[SupervisionSegment(..., start=0.2, duration=1.3, text="...")])` instead of a new format

File: examples/asr/asr_eou/README.md [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+
+
+### 2.7 Configuring dataset in model config
+
+Now we can update the model config to use the prepared training and evaluation data config.
+
+```yaml
+model:
+  train_ds:
+    rain_ds:
```
Comment (pzelasko) [inline examples/asr/asr_eou/README.md]: typo

File: scripts/asr_eou/transcribe_speech_sharded.py (hunk 7 above)
Comment (pzelasko) [reply scripts/asr_eou/transcribe_speech_sharded.py:43]: This script is a step in the right direction, but for real distributed, I feel we should add a proper support to NeMo Skills. 

I'm OK with whatever you decide for this PR guys, but I have some reservations against having both _parallel and _distributed variant in main branch, because people will start relying on both, and one (or both) are destined to disappear later...

File: nemo/collections/asr/data/audio_to_eou_label_lhotse.py [hunk 15]
```diff
@@ -0,0 +1,563 @@
...
+            if isinstance(c, MixedCut):
+                c = c.first_non_padding_cut
+
+            sample_ids.append(c.id)
+            audio_filepaths.append(c.recording.sources[0].source)
+
+            audio_i = audio[i]
+            audio_len_i = audio_lens[i]
+
+            # Maybe apply speed perturbation, this has to be done before getting the EOU labels
```
Comment (pzelasko) [inline nemo/collections/asr/data/audio_to_eou_label_lhotse.py]: What's the issue with speed perturbation in dataloader.py config? Can we fix that issue instead of adding new perturbation here?
BTW did you apply perturbation to EOU labels as well? cuts.perturb_speed would auto apply that to supervision segments

This one I am a little concerned about because changing the sample length after the Sampler invalidates bucketing decisions; if you combine this with OOMptimizer, not running into OOM is no longer guaranteed as some samples can be a little longer than expected in every bucket (especially longer data buckets since the absolute length delta is proportional to length itself).

File: nemo/collections/asr/data/audio_to_eou_label_lhotse.py [hunk 16]
```diff
@@ -0,0 +1,563 @@
...
+            audio_i, audio_len_i = self._maybe_augment_length(audio_i, audio_len_i)
+
+            # Get EOU labels and text tokens
+            eou_targets_i = self._get_frame_labels(c, audio_len_i)
+            text_tokens_i = self._get_text_tokens(c)
+
+            # Maybe apply random padding to both sides of the audio
+            audio_i, audio_len_i, eou_targets_i = self._random_pad_audio(audio_i, audio_len_i, eou_targets_i)
```
Comment (pzelasko) [inline nemo/collections/asr/data/audio_to_eou_label_lhotse.py:232]: is it possible to use `cut.pad()` (also auto adjusts supervision segment boundaries)?

File: nemo/collections/asr/data/audio_to_eou_label_lhotse.py [hunk 17]
```diff
@@ -0,0 +1,563 @@
...
+            audio_filepaths.append(c.recording.sources[0].source)
+
+            audio_i = audio[i]
+            audio_len_i = audio_lens[i]
+
+            # Maybe apply speed perturbation, this has to be done before getting the EOU labels
+            audio_i, audio_len_i = self._maybe_augment_length(audio_i, audio_len_i)
+
+            # Get EOU labels and text tokens
+            eou_targets_i = self._get_frame_labels(c, audio_len_i)
```
Comment (pzelasko) [inline nemo/collections/asr/data/audio_to_eou_label_lhotse.py:228]: is it possible to re-use `compute_supervisions_frame_mask(cut, frame_shift=...)` imported from `lhotse.cut.set`

**Review by pzelasko — COMMENTED** [review]
Great work!

File: scripts/asr_eou/transcribe_speech_sharded.py (hunk 7 above)
Comment (nithinraok) [reply scripts/asr_eou/transcribe_speech_sharded.py:43]: Suggestion: move the new script to the `asr_eou` folder so Steve’s work is preserved. It will still need tarred support and/or NeMo-skills integration. After that, it could replace `transcribe_speech_parallel.py`. Steve, let me know if this works for you.

Comment (stevehuang52) [reply scripts/asr_eou/transcribe_speech_sharded.py:43]: I'm renaming it to `transcribe_speech_sharded.py` so that the name is closer to what the script does. I'm moving it to the `scripts/asr_eou` folder.

File: nemo/collections/asr/data/audio_to_eou_label_lhotse.py (hunk 15 above)
Comment (stevehuang52) [reply nemo/collections/asr/data/audio_to_eou_label_lhotse.py]: The issue is the order of perturbation is not configurable in `dataloader.py`, and that the noise aug will have bugs if padding is applied before noise. 

I'm not actually using speed perturbation, but yes I didn't adjust the labels so I should remove speed perturb.

No I didn't use OOMptimzer since it doesn't actually run augmentations to estimate the batch sizes.

File: nemo/collections/asr/data/audio_to_eou_label_lhotse.py (hunk 17 above)
Comment (stevehuang52) [reply nemo/collections/asr/data/audio_to_eou_label_lhotse.py:228]: Thanks for the suggestion, I think it's a good choice for binary masks. Meanwhile, I would like to keep the current implementation for future extension where EOU labels are not binary (I actually have the code but it's not used in training the published model so I didn't include it in the PR)

File: nemo/collections/asr/data/audio_to_eou_label_lhotse.py (hunk 16 above)
Comment (stevehuang52) [reply nemo/collections/asr/data/audio_to_eou_label_lhotse.py:232]: Yeah I had a version that directly processes cut object instead of numpy array, but didn't find it to be improving speed so didn't dig further. Since most of the code here is handling numpy array, I think it's better to keep using numpy.

File: docker/Dockerfile.speech [hunk 18]
```diff
@@ -14,7 +14,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-ARG BASE_IMAGE=nvcr.io/nvidia/pytorch:24.07-py3
+ARG BASE_IMAGE=nvcr.io/nvidia/pytorch:25.12-py3
```
Comment (nithinraok) [inline docker/Dockerfile.speech:17]: If this PR merges after 26.02 release. Please update this.

**Review by nithinraok — APPROVED** [review]
LGTM
